src/analysis/history.cpp

```diff
@@ -279,4 +280,100 @@ void ReduceLocalizationFunction(MeshData<Real> *md) {
 
 } // exp
 
+using namespace parthenon::package::prelude;
+template <typename F>
+KOKKOS_INLINE_FUNCTION Real ComputeDivInPillbox(int ndim, int b, int k, int j, int i,
+                                                const parthenon::Coordinates_t &coords,
+                                                const F &f) {
+  Real div_mass_flux_integral;
+  div_mass_flux_integral =
+      -(f(b, 1, k, j, i + 1) - f(b, 1, k, j, i)) * coords.FaceArea<X1DIR>(k, j, i);
+
+  if (ndim >= 2) {
+    div_mass_flux_integral +=
+        -(f(b, 2, k, j + 1, i) - f(b, 2, k, j, i)) * coords.FaceArea<X2DIR>(k, j, i);
+  }
+  if (ndim >= 3) {
+    div_mass_flux_integral +=
+        -(f(b, 3, k + 1, j, i) - f(b, 3, k, j, i)) * coords.FaceArea<X3DIR>(k, j, i);
+  }
+  return div_mass_flux_integral;
+}
+
+// This function calculates mass accretion rate which I defined as
+// Mdot=Int(dV*d/dx^i{detg*rho*U^i}) where detg is the determinant of four metric, U is
+// four-velocity, and dV=d^3x
+
+Real CalculateMdot(MeshData<Real> *md, Real rc, bool gain) {
+  const auto ib = md->GetBoundsI(IndexDomain::interior);
+  const auto jb = md->GetBoundsJ(IndexDomain::interior);
+  const auto kb = md->GetBoundsK(IndexDomain::interior);
+  namespace c = fluid_cons;
+  using parthenon::MakePackDescriptor;
+  auto *pmb = md->GetParentPointer();
+  Mesh *pmesh = md->GetMeshPointer();
+  auto &resolved_pkgs = pmesh->resolved_packages;
+  const int ndim = pmesh->ndim;
+  using parthenon::PDOpt;
+  static auto desc = MakePackDescriptor<c::density, internal_variables::GcovHeat,
+                                        internal_variables::GcovCool>(
+      resolved_pkgs.get(), {}, {PDOpt::WithFluxes});
+  auto v = desc.GetPack(md);
+
+  const int nblocks = v.GetNBlocks();
+  auto geom = Geometry::GetCoordinateSystem(md);
+  Real result = 0.0;
+  const bool is_monopole_cart =
+      (typeid(PHOEBUS_GEOMETRY) == typeid(Geometry::MonopoleCart));
+  const bool is_monopole_sph =
+      (typeid(PHOEBUS_GEOMETRY) == typeid(Geometry::MonopoleSph));
+
+  parthenon::par_reduce(
+      parthenon::LoopPatternMDRange(), "Calculates mass accretion rate (SN)",
+      DevExecSpace(), 0, nblocks - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i, Real &lresult) {
+        const auto &coords = v.GetCoordinates(b);
+        const Real vol = coords.CellVolume(k, j, i);
+
+        Real C[NDFULL];
+        geom.Coords(CellLocation::Cent, b, k, j, i, C);
+        Real r = std::sqrt(C[1] * C[1] + C[2] * C[2] + C[3] * C[3]);
+        if (r <= rc) {
+          if (gain) {
+            auto rad = pmb->packages.Get("radiation").get();
+            const parthenon::AllReduce<bool> *pdo_gain_reducer =
+                rad->MutableParam<parthenon::AllReduce<bool>>("do_gain_reducer");
+            const bool do_gain = pdo_gain_reducer->val;
```

Comment:
This can't be here---you can't access packages from within a par for on GPU. Move this above.

Reply:
done